File: backend/src/services/image-downloader.ts
import fs from "node:fs";
import path from "node:path";
import { randomUUID } from "node:crypto";
import { lookup } from "node:dns/promises";
import net from "node:net";
import { config } from "../config.js";

const STORAGE_PATH = config.memeStoragePath;

// Ensure storage directory exists
fs.mkdirSync(STORAGE_PATH, { recursive: true });

function getExtension(contentType: string, url: string): string {
  const mimeMap: Record<string, string> = {
    "image/jpeg": ".jpg",
    "image/jpg": ".jpg",
    "image/png": ".png",
    "image/webp": ".webp",
    "image/gif": ".gif",
  };

  if (mimeMap[contentType]) return mimeMap[contentType];

  // Try to extract from URL path
  const urlPath = new URL(url).pathname;
  const ext = path.extname(urlPath);
  if (ext && [".jpg", ".jpeg", ".png", ".webp", ".gif"].includes(ext.toLowerCase())) {
    return ext.toLowerCase();
  }

  return ".jpg"; // default
}

export async function downloadImage(imageUrl: string): Promise<{ filePath: string; fileName: string }> {
  const url = parseSafeImageUrl(imageUrl);
  await assertHostnameResolvesPublicly(url.hostname);
  const response = await fetch(url, {
    signal: AbortSignal.timeout(config.imageDownloadTimeoutMs),
    redirect: "error",
  });

  if (!response.ok) {
    throw new Error(`Failed to download image: ${response.status} ${response.statusText}`);
  }

  const contentType = response.headers.get("content-type") || "image/jpeg";
  if (!contentType.toLowerCase().startsWith("image/")) {
    throw new Error(`Refusing to save non-image response: ${contentType}`);
  }

  const contentLength = Number(response.headers.get("content-length") || 0);
  if (contentLength > config.maxImageBytes) {
    throw new Error(`Image is too large: ${contentLength} bytes`);
  }

  const ext = getExtension(contentType, url.toString());
  const fileName = `${randomUUID()}${ext}`;
  const filePath = path.join(STORAGE_PATH, fileName);

  const buffer = Buffer.from(await response.arrayBuffer());
  if (buffer.length > config.maxImageBytes) {
    throw new Error(`Image is too large: ${buffer.length} bytes`);
  }

  fs.writeFileSync(filePath, buffer);

  console.log(`[MemeDrop] Image saved: ${filePath} (${buffer.length} bytes)`);

  return { filePath, fileName };
}

export function storedImagePathForPublicPath(publicPath: string): string | null {
  if (!publicPath.startsWith("/memes/")) return null;
  const fileName = path.basename(publicPath);
  if (!fileName || fileName !== publicPath.slice("/memes/".length)) return null;
  return path.join(STORAGE_PATH, fileName);
}

export async function deleteStoredImage(publicPath: string): Promise<boolean> {
  const filePath = storedImagePathForPublicPath(publicPath);
  if (!filePath) return false;

  try {
    await fs.promises.unlink(filePath);
    return true;
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") return false;
    throw err;
  }
}

function parseSafeImageUrl(imageUrl: string): URL {
  let url: URL;
  try {
    url = new URL(imageUrl);
  } catch {
    throw new Error("image_url must be an absolute URL");
  }

  if (!["http:", "https:"].includes(url.protocol)) {
    throw new Error("Only http(s) image URLs are allowed");
  }
  if (isBlockedHostname(url.hostname)) {
    throw new Error("Refusing to fetch local or private image URL");
  }

  return url;
}

function isBlockedHostname(hostname: string): boolean {
  const normalized = hostname.toLowerCase();
  if (["localhost", "0.0.0.0", "::1"].includes(normalized)) return true;
  if (isPrivateOrReservedIp(normalized)) return true;
  return false;
}

async function assertHostnameResolvesPublicly(hostname: string): Promise<void> {
  if (net.isIP(hostname)) {
    if (isPrivateOrReservedIp(hostname)) {
      throw new Error("Refusing to fetch local or private image URL");
    }
    return;
  }

  const addresses = await lookup(hostname, { all: true, verbatim: true });
  if (addresses.length === 0) {
    throw new Error("image_url hostname did not resolve");
  }

  const blocked = addresses.find((entry) => isPrivateOrReservedIp(entry.address));
  if (blocked) {
    throw new Error(`Refusing to fetch image URL that resolves to private IP ${blocked.address}`);
  }
}

export function isPrivateOrReservedIp(address: string): boolean {
  if (net.isIPv4(address)) return isPrivateOrReservedIpv4(address);
  if (net.isIPv6(address)) return isPrivateOrReservedIpv6(address);
  return false;
}

function isPrivateOrReservedIpv4(address: string): boolean {
  const parts = address.split(".").map((part) => Number(part));
  if (parts.length !== 4 || parts.some((part) => !Number.isInteger(part) || part < 0 || part > 255)) {
    return true;
  }
  const [a, b] = parts;

  if (a === 0) return true;
  if (a === 10) return true;
  if (a === 100 && b >= 64 && b <= 127) return true;
  if (a === 127) return true;
  if (a === 169 && b === 254) return true;
  if (a === 172 && b >= 16 && b <= 31) return true;
  if (a === 192 && b === 0) return true;
  if (a === 192 && b === 168) return true;
  if (a === 198 && (b === 18 || b === 19)) return true;
  if (a >= 224) return true;

  return false;
}

function isPrivateOrReservedIpv6(address: string): boolean {
  const normalized = address.toLowerCase();
  if (normalized === "::" || normalized === "::1") return true;
  if (normalized.startsWith("::ffff:")) {
    return isPrivateOrReservedIp(normalized.slice("::ffff:".length));
  }

  const first = Number.parseInt(normalized.split(":")[0] || "0", 16);
  if (!Number.isFinite(first)) return true;

  if ((first & 0xfe00) === 0xfc00) return true; // unique local fc00::/7
  if ((first & 0xffc0) === 0xfe80) return true; // link-local fe80::/10
  if ((first & 0xff00) === 0xff00) return true; // multicast ff00::/8
  if (normalized.startsWith("2001:db8:") || normalized === "2001:db8::") return true;

  return false;
}

File: apps/api/src/memedrop_api/services/image_downloader.py
from __future__ import annotations

import asyncio
import ipaddress
import mimetypes
import socket
from collections.abc import Awaitable, Callable
from pathlib import Path
from urllib.parse import urlsplit
from uuid import uuid4

import httpx

from memedrop_api.config import Settings

Resolver = Callable[[str], Awaitable[list[str]]]
MIME_EXTENSIONS = {
    "image/jpeg": ".jpg",
    "image/jpg": ".jpg",
    "image/png": ".png",
    "image/webp": ".webp",
    "image/gif": ".gif",
}


async def download_image(
    image_url: str,
    settings: Settings,
    *,
    client: httpx.AsyncClient | None = None,
    resolver: Resolver | None = None,
) -> tuple[Path, str]:
    parsed = urlsplit(image_url)
    if parsed.scheme not in {"http", "https"} or not parsed.hostname:
        raise ValueError("Only absolute http(s) image URLs are allowed")
    if parsed.username or parsed.password:
        raise ValueError("Image URLs must not contain credentials")
    await assert_hostname_resolves_publicly(parsed.hostname, resolver=resolver)

    owns_client = client is None
    request_client = client or httpx.AsyncClient(
        timeout=settings.image_download_timeout_ms / 1000,
        follow_redirects=False,
    )
    try:
        response = await request_client.get(image_url)
        response.raise_for_status()
    finally:
        if owns_client:
            await request_client.aclose()

    content_type = response.headers.get("content-type", "image/jpeg").split(";", 1)[0].lower()
    if not content_type.startswith("image/"):
        raise ValueError(f"Refusing to save non-image response: {content_type}")
    declared_size = int(response.headers.get("content-length", "0") or 0)
    if declared_size > settings.max_image_bytes or len(response.content) > settings.max_image_bytes:
        raise ValueError("Image is too large")

    extension = MIME_EXTENSIONS.get(content_type)
    if extension is None:
        guessed = Path(parsed.path).suffix.lower()
        extension = (
            guessed
            if guessed in {".jpg", ".jpeg", ".png", ".webp", ".gif"}
            else mimetypes.guess_extension(content_type) or ".jpg"
        )
    file_name = f"{uuid4()}{extension}"
    file_path = settings.meme_storage_path / file_name
    await asyncio.to_thread(file_path.parent.mkdir, parents=True, exist_ok=True)
    await asyncio.to_thread(file_path.write_bytes, response.content)
    return file_path, file_name


async def resolve_hostname(hostname: str) -> list[str]:
    entries = await asyncio.to_thread(
        socket.getaddrinfo, hostname, None, socket.AF_UNSPEC, socket.SOCK_STREAM
    )
    return list({str(entry[4][0]) for entry in entries})


async def assert_hostname_resolves_publicly(
    hostname: str, *, resolver: Resolver | None = None
) -> None:
    if hostname.lower() == "localhost":
        raise ValueError("Refusing to fetch local or private image URL")
    try:
        literal = ipaddress.ip_address(hostname)
        addresses = [str(literal)]
    except ValueError:
        addresses = await (resolver or resolve_hostname)(hostname)
    if not addresses:
        raise ValueError("image_url hostname did not resolve")
    for address in addresses:
        if is_private_or_reserved_ip(address):
            raise ValueError(f"Refusing to fetch image URL that resolves to private IP {address}")


def is_private_or_reserved_ip(address: str) -> bool:
    try:
        parsed = ipaddress.ip_address(address)
        return not parsed.is_global or parsed.is_multicast
    except ValueError:
        return True


def stored_image_path(public_path: str, storage_path: Path) -> Path | None:
    prefix = "/memes/"
    if not public_path.startswith(prefix):
        return None
    file_name = public_path.removeprefix(prefix)
    if not file_name or Path(file_name).name != file_name:
        return None
    return storage_path / file_name


async def delete_stored_image(public_path: str, storage_path: Path) -> bool:
    file_path = stored_image_path(public_path, storage_path)
    if file_path is None:
        return False
    try:
        await asyncio.to_thread(file_path.unlink)
        return True
    except FileNotFoundError:
        return False
